sync/sync_tail.go

```diff
@@ -0,0 +1,165 @@
+package sync
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/celestiaorg/go-header"
+)
+
+// TODO:
+//  * Refactor tests
+//  * Ensure sync always happen on start
+
+// subjectiveTail returns the current Tail header.
+// Lazily fetching it if it doesn't exist locally or moving it to a different height.
+// Moving is done if either parameters are changed or tail moved outside a pruning window.
+func (s *Syncer[H]) subjectiveTail(ctx context.Context, head H) (H, error) {
+	tail, err := s.store.Tail(ctx)
+	if err != nil && !errors.Is(err, header.ErrEmptyStore) {
+		return tail, err
+	}
+
+	var fetched bool
+	if tailHash, outdated := s.isTailHashOutdated(tail); outdated {
+		log.Debugw("tail hash updated", "hash", tailHash)
+		tail, err = s.store.Get(ctx, tailHash)
+		if err != nil {
+			log.Debugw("tail hash not available locally, fetching...", "hash", tailHash)
+			tail, err = s.getter.Get(ctx, tailHash)
+			if err != nil {
+				return tail, fmt.Errorf("getting SyncFromHash tail(%x): %w", tailHash, err)
+			}
+			fetched = true
+		}
+	} else if tailHeight, outdated := s.isTailHeightOutdated(tail); outdated {
+		log.Debugw("tail height updated", "height", tailHeight)
+		if tailHeight <= s.store.Height() {
+			tail, err = s.store.GetByHeight(ctx, tailHeight)
+		}
+		if err != nil || tailHeight != tail.Height() {
+			log.Debugw("tail height not available locally, fetching...", "height", tailHeight)
+			tail, err = s.getter.GetByHeight(ctx, tailHeight)
+			if err != nil {
+				return tail, fmt.Errorf("getting SyncFromHeight tail(%d): %w", tailHeight, err)
+			}
+			fetched = true
+		}
+	} else if tailHash == nil && tailHeight == 0 {
+		if tail.IsZero() {
+			// no previously known Tail available - estimate solely on Head
```

Comment:
let's add a debug log for this case as well pls that includes estimated height

Reply:
it's debugged after refactoring in a different place